crates/executor/src/executor.rs

```diff
@@ -17,258 +19,423 @@ use revm::{
-pub enum AccountInfoChangeSet {
-    /// The account is newly created.
-    Created {
-        /// The newly created account.
-        new: Account,
-    },
-    /// An account was deleted (selfdestructed) or we have touched
-    /// an empty account and we need to remove/destroy it.
-    /// (Look at state clearing [EIP-158](https://eips.ethereum.org/EIPS/eip-158))
-    Destroyed {
-        /// The account that was destroyed.
-        old: Account,
-    },
-    /// The account was changed.
-    Changed {
-        /// The account after the change.
-        new: Account,
-        /// The account prior to the change.
-        old: Account,
-    },
-    /// Nothing was changed for the account (nonce/balance).
-    NoChange,
+pub struct Executor<'a, DB>
+where
+    DB: StateProvider,
+{
+    chain_spec: &'a ChainSpec,
+    evm: EVM<&'a mut SubState<DB>>,
 }
 
-impl AccountInfoChangeSet {
-    /// Apply the changes from the changeset to a database transaction.
-    pub fn apply_to_db<'a, TX: DbTxMut<'a>>(
-        self,
-        tx: &TX,
-        address: Address,
-        tx_index: u64,
-        has_state_clear_eip: bool,
-    ) -> Result<(), DbError> {
-        match self {
-            AccountInfoChangeSet::Changed { old, new } => {
-                // insert old account in AccountChangeSet
-                // check for old != new was already done
-                tx.put::<tables::AccountChangeSet>(
-                    tx_index,
-                    AccountBeforeTx { address, info: Some(old) },
-                )?;
-                tx.put::<tables::PlainAccountState>(address, new)?;
-            }
-            AccountInfoChangeSet::Created { new } => {
-                // Ignore account that are created empty and state clear (SpuriousDragon) hardfork
-                // is activated.
-                if has_state_clear_eip && new.is_empty() {
-                    return Ok(())
-                }
-                tx.put::<tables::AccountChangeSet>(
-                    tx_index,
-                    AccountBeforeTx { address, info: None },
-                )?;
-                tx.put::<tables::PlainAccountState>(address, new)?;
-            }
-            AccountInfoChangeSet::Destroyed { old } => {
-                tx.delete::<tables::PlainAccountState>(address, None)?;
-                tx.put::<tables::AccountChangeSet>(
-                    tx_index,
-                    AccountBeforeTx { address, info: Some(old) },
-                )?;
-            }
-            AccountInfoChangeSet::NoChange => {
-                // do nothing storage account didn't change
+impl<'a, DB> Executor<'a, DB>
+where
+    DB: StateProvider,
+{
+    fn new(chain_spec: &'a ChainSpec, db: &'a mut SubState<DB>) -> Self {
+        let mut evm = EVM::new();
+        evm.database(db);
+        Executor { chain_spec, evm }
+    }
+
+    fn db(&mut self) -> &mut SubState<DB> {
+        self.evm.db().expect("db to not be moved")
+    }
+
+    fn recover_senders(
```

Comment:
This is a bit weird because it is "if there are no senders specified, try to get them from the block bodies", but I guess it's fine because downstream consumers will call execute with `senders = None` and it'll cover for this internally.

Reply:
Yes, I also think it looks weird. But considering performance for `stages`, I decided to implement it like this way because I couldn't find a better way. Please improve and revise it then I'll learn from the future works.

Reply:
Yeah no strong feelings.